File: #0057.insert-interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {

        int size = intervals.size();
        int index = 0;
        vector<vector<int>> result;
        while (index < size && intervals[index][1] < newInterval[0]) {
            result.push_back(intervals[index]);
            ++index;
        }

        int start = newInterval[0];
        int end = newInterval[1];
        while (index < size && intervals[index][0] <= end) {
            start = min(intervals[index][0], start);
            end = max(intervals[index][1], end);
            ++index;
        }

        result.push_back({start, end});
        
        while (index < size) {
            result.push_back(intervals[index]);
            ++index;
        }

        return result;
    }
};

File: #0057.insert-interval.js
/**
 * @param {number[][]} intervals
 * @param {number[]} newInterval
 * @return {number[][]}
 */
var insert = function(intervals, newInterval) {
    intervals.push(newInterval);
    intervals.sort((a, b) => a[0] - b[0]);

    for (let i = 1; i < intervals.length; i++) {
        const prev = intervals[i - 1];
        const cur = intervals[i];

        if (prev[1] >= cur[0]) {
            intervals[i] = [prev[0], Math.max(prev[1], cur[1])];
            intervals.splice(i - 1, 1);
            i--;
        }
    }
    return intervals;
};
